### ISSUE TITLE
question about warmup

### ISSUE THREAD

**Issue by ggslayer** [triton-inference-server/server#3696]
Hi, I'm using triton r21.10, gpu, I use tensorflow1 and onnxruntime backends, make a ensemble model to pipeline the onnx and pb models above, it runs ok!

But, I notice that, run my test program to call infer continuously, the throughput reached max after about 4 minites later... , at begin of time,  the throughtput is 1/s only , and gpu usage is 82%, and after 10 seconds, the throughtput is 2/s, the gpu usage is also 82%, the throughtput grows slowly slowly..  after 4 minites, it grow up to 21/s, the max value I excepted.

When I run onnxruntime to infer the onnx model, the warnup is quick.
I didn't test the pb model separately. (the model is old model used many months, i think there is no problem)

Is this normal? or something wrong?  please give me some expirements about this, thank you.

**Comment by Tabrizian** [comment]
Is your Python model using any framework (e.g. PyTorch or Tensorflow)? If not, then I also do not expect any warmup required for the Python model. 

It also depends on how the requests are generated by your test program. Does it wait for the execution of the requests to be finished before starting the next request or it keeps generating requests and does not wait for its response? If your models have multiple instances and are using Execution Providers I think it may take sometime until the model is warmed up.

Using model warmup settings, you can also make sure that your models are properly warmed up before they are ready to execute the requests. https://github.com/triton-inference-server/server/blob/r21.11/docs/model_configuration.md#model-warmup

cc @tanmayv25

**Comment by tanmayv25** [comment]
Can you describe more about your ensemble pipeline? Is it just model A -> model B ? I would expect large latency only for the first few request. However, in your case the setup doesn't get warmed up till atleast 240 requests..  As @Tabrizian described, multiple instances can take large time to warmup. That is warming up model A and model B might not be large in separation. But can be large if using special pipeline. For instance, model A feeding data to multiple model Bs.

**Comment by ggslayer** [comment]
there is only one instance for the two models,  the first model is onnx, second is tensorflow1,  in the ensemble config , the onnx output is transfer to the input of the second model just like this:

```

name: "demo"
platform: "ensemble"
max_batch_size: 0

input [
  {
      name: "demo_input1"
      data_type: TYPE_INT64
      dims: [ -1 ]
    },
    {
      name: "demo_input2"
      data_type: TYPE_FP32
      dims: [ -1, -1 ]
    }
]

output [
  {
    name: "demo_output"
    data_type: TYPE_FP32
    dims: [ -1,-1 ]
  }
]

ensemble_scheduling {
  step [
    {
      model_name: "model_onnx"
      model_version: -1
      input_map {
        key: "onnx_input1"
        value: "demo_input1"
      }
      input_map {
        key: "onnx_input2"
        value: "demo_input2"
      }
      
      output_map {
        key: "onnx_output"
        value: "middle_output"
      }
    },
    {
      model_name: "model_tensorflow"
      model_version: -1
      input_map {
        key: "tf_input1"
        value: "middle_output"
      }
      output_map {
        key: "tf_output"
        value: "demo_output"
      }
    }
  ]
}

response_cache { 
  enable: True 
}
```

**Comment by tanmayv25** [comment]
Some follow-up questions:
1.  I see you are using dynamic shapes. Are all your requests with same shape tensors and values?
2. You seem to be using response cache. It is possible that actual latency of the model is indeed 1sec. However, the response cache comes into the picture later and reduce the latency. Can you try running experiment without response cache? 
3. Can you share the configs of composing models as well? 
4. The output from a [perf_analyzer ](https://github.com/triton-inference-server/server/blob/main/docs/perf_analyzer.md) run will also help us in understanding more about where the inference request is spending most of its time. Please run perf_analyzer for 3 different model configs. 
     a. model_tensorflow
     b. model_onnx
     c. demo

**Comment by handoku** [comment]
To my knowledge, using tf model with dynamic shape is a disaster, tensorflow runtime need to lazy initialize computation graph for every single unique shape, which result in very high inference latency.

I find zero method to solve this problem yet, I will appreciate it if anyone find a solution and share with us.

Right now, my suggestions is:

1. Using fixed shape
2. Before you start your real request, enum every single unique shape to do a brute-force warmup.

**Comment by dyastremsky** [comment]
Closing issue due to lack of activity. Please re-open the issue if you would like to follow up.

**Comment by saeid93** [comment]
@Tabrizian For a use case I need to measure the warm-up latency of each model (the time it takes for each model to get to its expected throughput rate, like 21/s in the original question of this issue), is there a way to do that? For example, do I need to continuously send requests to the model until the r/s get to their peak, or there is a cleaner way to do this?

**Comment by Tabrizian** [comment]
There is no automated way of doing this in Triton right now. You can specify a large enough `count` in the model warmup setting to make sure your model is fully warmed up. https://github.com/triton-inference-server/common/blob/main/protobuf/model_config.proto#L1681
